Closes #2395: Re-vamp of get-started page (#2402)
pharmaverse/admiral: vignettes/admiral.Rmd

Author: implemented

---
title: "Simple ADSL & ADVS Walkthrough"
output: rmarkdown::html_vignette
vignette: >

>   %\VignetteIndexEntry{Simple ADSL & ADVS Walkthrough}

Author: i had missed that i needed to change it back - fixed.

# Main Idea

The main idea of `{admiral}` is that an ADaM dataset is built by a sequence of
derivations. Each derivation adds one or more variables or records to the
processed dataset. This modular approach makes it easy to adjust code by adding,
removing, or modifying derivations. Each derivation is a function call. 

In this vignette we will explore some of the different type of derivation functions 
offered by `{admiral}`, as well as the argument conventions they follow and how best 
to start an `{admiral}` script. To help with the examples showcased, we will load 

> some packages and set up some example data to manipulate later. The `{dplyr}` and 
> `{lubridate}` and `{stringr}` packages are `{tidyverse}` packages and are used heavily

Reviewer: ```suggestion
some packages and set up some example data to manipulate later. The `{dplyr}`,
`{lubridate}` and `{stringr}` packages are `{tidyverse}` packages and are used heavily
```

throughout this vignette.  The `{admiral}` package also leverages the `{pharmaversesdtm}` 
package for example SDTM datasets which are from the CDISC Pilot Study.

```{r, message=FALSE, warning=FALSE}

> library(dplyr, warn.conflicts = FALSE)

Reviewer: Do we want a commented out install.packages(c("dplyr", .... , "admiral") so a user could quickly install every package ?

library(lubridate)
library(stringr)
library(tibble)
library(pharmaversesdtm)
library(admiral)

# Read in SDTM datasets
data("dm")
data("ds")
data("ex")
data("vs")
data("admiral_adsl")

dm <- convert_blanks_to_na(dm)
ds <- convert_blanks_to_na(ds)
ex <- convert_blanks_to_na(ex)
vs <- convert_blanks_to_na(vs)

> ex_ext <- ex %>%

Reviewer: What about breaking this up a bit?  Load and say why we convert blanks to NA - a huge stuggle for new users

then move into the derive functions
![image](https://github.com/pharmaverse/admiral/assets/10111024/d34f8622-4938-4f9f-9246-2ddbda28c667)

Reviewer: what about a little note about us preparing datasets to showcase adding variables for ADSL and adding records for ADVS?

Reviewer 2: I would remove the `convert_blanks_to_na()` chunk. It is not necessary for pharmaversesdtm datasets and only necessary for hybrid workflows where the SDTMs are created with SAS and the ADaMs with R. Thus we don't need to present it on the "Get Started" page.

Author: happy to remove for simplicity, vignette is already getting quite long

processed dataset. This modular approach makes it easy to adjust code by adding,
removing, or modifying derivations. Each derivation is a function call. 

In this vignette we will explore some of the different type of derivation functions 
offered by `{admiral}`, as well as the argument conventions they follow and how best 
to start an `{admiral}` script. 

# Setup

To help with the examples showcased, we will load some packages and set up some example 
data to manipulate later. The `{dplyr}`, `{lubridate}` and `{stringr}` packages are 
`{tidyverse}` packages and are used heavily throughout this vignette. The `{admiral}`
package also leverages the `{pharmaversesdtm}` package for example SDTM datasets 
which are from the CDISC Pilot Study.

```{r, message=FALSE, warning=FALSE}
# Uncomment line below if you need to install these packages
# install.packages(c("dplyr", "lubridate", "stringr", "tibble", "pharmaversesdtm", "admiral"))

library(dplyr, warn.conflicts = FALSE)
library(lubridate)
library(stringr)
library(tibble)
library(pharmaversesdtm)
library(admiral)

# Read in SDTM datasets
data("dm")
data("ds")
data("ex")
data("vs")
data("admiral_adsl")
```

> The `adsl` and `advs` objects will now be prepared to showcase [addition of variables](#addvars)

Reviewer 3: Can we add a note to refer to the Handling missing data section ? I found it useful to explain on the "Get started" vignette why we have to convert the blanks to NA: the {admiral} functions don't do it automatically. I think it's one of the basics to know, especially if users are also new to R (not only to {admiral}) it was a good reminder for a good start.

Author: @Fanny-Gautier added a note

Reviewer 3: Nice, thanks

derivations. Each derivation adds one or more variables or parameters to the

> processed dataset. This modular approach makes it easy to adjust code by adding,

Reviewer: Could we say records instead of parameters?

Reviewer 3: done

Functions which derive a dedicated parameter start with `derive_param_` followed by
the parameter name, e.g., `derive_param_bmi()` derives the `BMI` parameter.

## Example: Adding Variables

Below we can see an example call to one of the most common derivation functions,
`derive_vars_merged()`. This function adds variable(s) to the input dataset based on the 
contents of another dataset. In this example, we add the treatment start datetime and 

> corresponding imputation flag (`EXSTTMF`) to `adsl` by identifying the first record

Reviewer: ```suggestion
corresponding imputation flag (`EXSTTMF`) to `ADSL` by identifying the first record
```
I think we do upper case for variables and datasets

Author: here i was referencing the actual `adsl` object in my code rather than the general `ADSL` dataset

Reviewer: oh i see.  okay - do we do this in other ADAM vignettes??  I guess it is no biggie

    order = exprs(EXSTDTM, EXSEQ),
    mode = "first",
    by_vars = exprs(STUDYID, USUBJID)
  )
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  adsl,
  display_vars = exprs(USUBJID, TRTSDTM, TRTSTMF)
)
```

## Example: Adding Observations

Another common derivation function is `derive_param_computed()`. This function
adds a derived parameter to an input dataset. In the example below, we use it
to derive the Mean Arterial Pressure (MAP) from the Systolic and Diastolic values
of the blood pressure. The parameters that are needed for the derivation are 
specified in the `parameters` argument, and within `set_values_to` we set all 
the variable values for the new derived record.

```{r eval=TRUE}
advs <- advs %>%
  derive_param_computed(
    by_vars = exprs(USUBJID, AVISIT, AVISITN),
    parameters = c("SYSBP", "DIABP"),
    set_values_to = exprs(
      AVAL = (AVAL.SYSBP + 2 * AVAL.DIABP) / 3,
      PARAMCD = "MAP",
      PARAM = "Mean Arterial Pressure (mmHg)",
      AVALU = "mmHg"
    )
  )
```

```{r, eval=TRUE, echo=FALSE}
advs %>%
  arrange(USUBJID, AVISITN, PARAMCD) %>%
  dataset_vignette(
    display_vars = exprs(USUBJID, AVISIT, PARAMCD, AVAL)
  )
```

> Note: For the users' convenience, `{admiral}` actually provides

Reviewer: ```suggestion
**Note:** For the users' convenience, `{admiral}` actually provides
```

`derive_param_map()`  (wrapper of `derive_param_computed()`) to derive MAP.
The above example serves for illustrative purposes only.

# Other Types of Functions

> Along with the derivation functions, `{admiral}` provides a large arsenal

Reviewer: A large arsenal? 
![image](https://github.com/pharmaverse/admiral/assets/10111024/909f058e-3c90-49aa-aa12-b37df652f9ed)

of functions to help you with your ADaM derivations. Here are some of the 
other categories.

## Higher Order Functions

[Higher order functions](../reference/index.html#higher-order) 
are `{admiral}` functions that take other functions as input. 
They enhance the existing portfolio of derivation functions by allowing greater 
customization of the latter's behavior. This is done by allowing a derivation
function to be:

> * Called multiple times, while varying some of the input arguments (`call_derivation()`);

Reviewer: ```suggestion
* `call_derivation()` - Called multiple times, while varying some of the input arguments.
```

> * Executed on a subset of the input dataset (`restrict_derivation()`);
> * Executed differently on subsets of the input dataset (`slice_derivation()`).

Reviewer: ```suggestion
* `restrict_derivation()` - Executed on a subset of the input dataset.
* `slice_derivation()` - Executed differently on subsets of the input dataset.
```

Higher order functions are a relatively advanced topic within `{admiral}`; you can

> can read all about them in their [dedicated vignette](higher_order.html).

Reviewer: SHould we just say the name of the vignette?

Author: changed

> can read all about them in their [dedicated vignette](../reference/index.html#higher_order).

Reviewer 2: ```suggestion
can read all about them in their [dedicated vignette](higher_order.html).
```

## Computation Functions

[Computations](../reference/index.html#computation-functions-for-vectors) expect vectors as
input and return a vector. Usually these computation functions can not be used with `%>%`. 
These functions can be used in expressions like `convert_dtc_to_dt()` in the derivation
of the Final Lab Visit Date (`FINLABDT`) in the example below:

> ```{r, eval=TRUE}

Reviewer: What about mentioning that you can use them in a `mutate()` functions as well?  Example?

Author: added

```{r, eval = TRUE}
advs_lastmap <- advs %>%
  filter(PARAMCD == "MAP") %>%
  filter_extreme(
    by_vars = exprs(USUBJID),
    order = exprs(AVISITN, PARAMCD),
    mode = "last"
  )
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  advs_lastmap,
  display_vars = exprs(USUBJID, AVISIT, PARAMCD, AVAL)
)
```

# Argument Conventions 

Within the `{admiral}` package, any arguments which expect variable names or expressions of variable names, symbols or 
expressions must be specified rather than strings.

- For arguments which expect a single variable name, the name can be specified without quotes and quotation, e.g.
`new_var = TEMPBL`

- For arguments which expect one or more variable names, a list of symbols is expected, e.g. `by_vars = exprs(PARAMCD, AVISIT)` 

- For arguments which expect a single expression, the expression needs to be passed "as is", e.g. `filter = PARAMCD == "TEMP"`

- For arguments which expect one or more expressions, a list of expressions is expected, e.g. `order = exprs(AVISIT, desc(AESEV))`

If you are new to expressions, consider reading our [Introduction to expressions](concepts_conventions.html#exprs) to

> learn more.

Reviewer: Should we just use the name of the vignette section?

Author: changed

can also be used on their own to explore or manipulate a dataset. For instance, in the example 
below we use `filter_extreme()` to extract the most recent MAP records in `advs`:

```{r, eval = TRUE}
advs_lastmap <- advs %>%
  filter(PARAMCD == "MAP") %>%
  filter_extreme(
    by_vars = exprs(USUBJID),
    order = exprs(AVISITN, PARAMCD),
    mode = "last"
  )

advs_lastmap %>%
  select(USUBJID, AVISIT, PARAMCD, AVAL)
```
# Argument Conventions 

Within the `{admiral}` package, any arguments which expect variable names or expressions of variable names, symbols or 
expressions must be specified rather than strings.

- For arguments which expect a single variable name, the name can be specified without quotes and quotation, e.g.
`new_var = TEMPBL`

- For arguments which expect one or more variable names, a list of symbols is expected, e.g. `by_vars = exprs(PARAMCD, AVISIT)` 

- For arguments which expect a single expression, the expression needs to be passed "as is", e.g. `filter = PARAMCD == "TEMP"`

- For arguments which expect one or more expressions, a list of expressions is expected, e.g. `order = exprs(AVISIT, desc(AESEV))`

> If you are new to expressions, consider reading our [Introduction to expressions](#../reference/index.html#concepts_conventions) to

Reviewer 2: ```suggestion
If you are new to expressions, consider reading our [Introduction to expressions](concepts_conventions.html#exprs) to
```

```

At this stage, only `PARAMCD` is required to perform the derivations. Additional
derived parameters may be added, so only `PARAMCD` is joined to the datasets at
this point. All other variables related to `PARAMCD` (e.g. `PARAM`, `PARAMCAT1`, ...)
will be added when all `PARAMCD` are derived.

```{r eval=TRUE}
advs <- derive_vars_merged_lookup(
  advs,
  dataset_add = param_lookup,
  new_vars = exprs(PARAMCD),
  by_vars = exprs(VSTESTCD)
)
```

```{r, eval=TRUE, echo=FALSE}
advs_param <- distinct(advs, USUBJID, PARAMCD, VSTESTCD)

dataset_vignette(advs_param, display_vars = exprs(USUBJID, VSTESTCD, PARAMCD))
```

## Derive Results (`AVAL`, `AVALC`)

The mapping of `AVAL` and `AVALC` is left to the ADaM programmer. An 
example mapping may be:

```{r eval=TRUE}
advs <- mutate(
  advs,
  AVAL = VSSTRESN,
  AVALC = VSSTRESC
)
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  advs,
  display_vars = exprs(VSTESTCD, PARAMCD, VSSTRESN, VSSTRESC, AVAL, AVALC),
  filter = USUBJID == "01-716-1024"
)
```

## Derive `MAP` Parameter

Below is finally our example of creating `Mean Arterial Pressure` for `ADVS` using
the functions `derive_param_map`.

> Note that only variables specified in the `by_vars` argument will be populated in the newly 

Reviewer: Can we mention that this is a wrapper around derive_param_computed()

```r
use_ad_template(
adam_name = "adsl",

> save_path = "./ad_adsl.R"

Reviewer 2: Why has the indentation be removed?

Author: accident - fixed

)
```

A list of all available templates can be obtained by `list_all_templates()`:

```{r}
library(admiral)
list_all_templates()
```

# Support

Support is provided via the [admiral Slack
channel](https://app.slack.com/client/T028PB489D3/C02M8KN8269). Additionally,
please feel free to raise issues in our [GitHub repository](https://github.com/pharmaverse/admiral/issues).

# See also

> 

Reviewer 2: Should we add a link to the "Programming Concepts and Conventions" vignette?

Author: good suggestion, added

A list of all available templates can be obtained by `list_all_templates()`:

```{r}

> library(admiral)

Author: amended
